### PR TITLE
replace Seq methods to work with a concrete type

### PR DESCRIPTION
**Pull request by cryptix**
This has been brewing in me for quite a while. This change

1) renames the interface `margaret.Seq` to `margaret.Seqer`
2) removes the concrete numerical type `margaret.BaseSeq` and replaces it with `int64`
3) changes `Seq() luigi.Observable` to `Seq() int64` in `margarte.Log`
4) adds `Changes() luigi.Observable`  to `margaret.Log` for the original feature of registering to changes via a callback.

With this change it's now possible to get the current sequence number of a log with a single call, as an concrete type, without type asserting it out of a `lugi.Observable`.

```go
        // Before:
	currVal, err := log.Seq().Value() // access observable and get the current value 
        r.NoError(err, "failed to get current sequence")
        //  type assert to the seq interface to getthe numerical value
        currSeq := currSeq.(margaret.Seq).Seq() 
        assert.Equal(t, margaret.BaseSeq(23), currSeq)

        // After 
	currSeq := log.Seq()
        assert.Equal(t, int64(23), currSeq)
```

**Outcome:**
`func Get(s int64) ...` and `func Append(...) (int64, error)` now pass and return a concrete typed, not interface values, which apart from being nicer to work with, should also remove some GC strain.

This also removes the need for a bunch of extra `.Seq()` function calls (or asserting to the concrete type) all the time. 

**Background:** initially we thought we might want to work with multiple layers of sequences (receive > indexed >logical) and thought the interface gives us more freedom to work on that but the concrete need for this never materialized.

ps: I only found one instance in go-ssb that uses the `Changes()` feature/behavior but lots and lots of cases using the `.Seq().Value()` construction.

pps: keeping `margaret.Seq` around just as an alias for `int64` seemed bad. Also people who want to implement `margarte.Seqer` need to import `margaret` which seemed bad to.

### CODE REVIEW

File: indexes/badger/index.go [hunk 1]
```diff
@@ -94,7 +95,7 @@ func newIndex(db *badger.DB, tipe interface{}, keyPrefix []byte) indexes.SeqSett
 		db:     db,
 		tipe:   tipe,
 		obvs:   make(map[indexes.Addr]luigi.Observable),
-		curSeq: margaret.BaseSeq(-2),
+		curSeq: int64(-2),
```
Comment (cblgh) [inline indexes/badger/index.go:98]: even this small diff shows how much cleaner, conceptually, this pr is. new people no longer need to spend time or working memory slots considering _"what is a `BaseSeq`?"_

:100:

File: legacyflumeoffset/query.go [hunk 2]
```diff
@@ -100,11 +100,11 @@ func (qry *lfoQuery) Reverse(yes bool) error {
 // 	if err != nil {
 // 		return errors.Wrap(err, "setCursorToLast: failed to establish current value")
 // 	}
-// 	currSeq, ok := v.(margaret.Seq)
```
Comment (cblgh) [inline legacyflumeoffset/query.go:103]: remove or comment why `setCursorToLast()` exists as a commented out block? maybe linking to an issue, if it's commented

File: legacyflumeoffset/test/simple_test.go [hunk 3]
```diff
@@ -44,9 +44,9 @@ func TestSimple(t *testing.T) {
 			for i, v := range tc.values {
 				seq, err := log.Append(v)
 				r.NoError(err, "error appending to log")
-				//r.Equal(margaret.BaseSeq(i), seq, "sequence missmatch")
-				r.True(int(seq.Seq()) >= i, "increasing sequence")
-				t.Log("entry", i, "seq:", seq.Seq())
+				//r.EqualValues(i, seq, "sequence missmatch")
```
Comment (cblgh) [inline legacyflumeoffset/test/simple_test.go:47]:
```suggestion
```

File: mem/log.go [hunk 4]
```diff
@@ -147,11 +151,11 @@ func (log *memlog) Query(specs ...margaret.QuerySpec) (luigi.Source, error) {
 	return qry, nil
 }
 
-func (log *memlog) Append(v interface{}) (margaret.Seq, error) {
+func (log *memlog) Append(v interface{}) (int64, error) {
 	log.l.Lock()
 	defer log.l.Unlock()
 	if log.closed {
-		return nil, io.ErrClosedPipe // already closed
+		return -2, io.ErrClosedPipe // already closed
```
Comment (cblgh) [inline mem/log.go:158]: why -2 btw? seen this a couple of times now

Comment (cblgh) [reply mem/log.go:158]: is this equivalent to `margaret.SeqSublogDeleted` that i've seen elsewhere?

File: multilog/test/sink.go [hunk 5]
```diff
@@ -77,8 +77,8 @@ func SinkTestSimple(f NewLogFunc) func(*testing.T) {
 
 			// append values
 			for i, v := range tc.values {
-				//err := sink.Pour(ctx, multilog.WithValue(margaret.BaseSeq(i), v))
-				err := sink.Pour(ctx, margaret.WrapWithSeq(v, margaret.BaseSeq(i)))
+				//err := sink.Pour(ctx, multilog.WithValue(int64(i), v))
```
Comment (cblgh) [inline multilog/test/sink.go:80]:
```suggestion
```

File: offset2/alter_replace_test.go [hunk 6]
```diff
@@ -12,7 +12,6 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.cryptoscope.co/luigi"
-	"go.cryptoscope.co/margaret"
```
Comment (cblgh) [inline offset2/alter_replace_test.go:15]: noice :)

File: offset2/qry.go [hunk 7]
```diff
@@ -92,15 +92,7 @@ func (qry *offsetQuery) Reverse(yes bool) error {
...
-	v, err := qry.log.seq.Value()
-	if err != nil {
-		return fmt.Errorf("setCursorToLast: failed to establish current value: %w", err)
-	}
-	currSeq, ok := v.(margaret.Seq)
-	if !ok {
-		return fmt.Errorf("setCursorToLast: failed to establish current value")
-	}
-	qry.nextSeq = margaret.BaseSeq(currSeq.Seq())
+	qry.nextSeq = qry.log.seqCurrent
```
Comment (cblgh) [inline offset2/qry.go:95]: what a quality of life improvement, lol

**Review by cblgh — APPROVED** [review]
yess!!! a great improvement & implementation :)

File: mem/log.go (hunk 4 above)
Comment (cblgh) [reply mem/log.go:158]: and if not: does a constant make sense to introduce? lotta -2 scattered around ^^

Comment (cryptix) [reply mem/log.go:158]: agreed, there should be a defined constant here.

`read()` and `write()` return negative counts when the call fails, that's the analogy here at least.

butttt `-1` is _empty log_ so `-2` should mean something like _something went wrong_. Good call, i will define it.

**Pull request merged**